enhancements/oc/mirroring-release-image-signatures.md

```diff
@@ -0,0 +1,256 @@
+Changing to previous `oc` may result in the loss of functionalty, as has been the case for other `oc` features, but the failures will be "option `<whatever>` not recognized" fast-failures, not silent behavior changes.
+
+### Version Skew Strategy
+
+Besides the user <-> `oc` interaction discussed in the previous section, we should also consider the `oc` <-> cluster interaction.
+
+If the user invokes an older `oc`, it may not create the release image signature ConfigMap expected by a newer cluster.
+Even for matching versions, users using [the *pushing to disk* flow](#pushing-to-disk) may decide not to apply the release image signature ConfigMap manifest to the target cluster.
+Cluster components that consume the release image signature ConfigMap should be robust in their absence.
+Obviously, any behavior which depends on missing objects will not happen.
+For example, the cluster-version operator will not be able to retrieve signatures from release image signature ConfigMaps if those ConfigMaps do not exist.
+
+If the user invokes a newer `oc`, it may create additional configuration manifests not supported by an older cluster.
+For example, a 4.5 `oc`, implementing this enhancement, might be used to [apply the release image signature ConfigMap](#applying-directly-to-the-target-cluster) to a 4.2 cluster whose cluster-version operator does not load signatures from ConfigMaps.
+Users in that situation might be surprised that the cluster was unable to locate release image signatures, not realizing that the older cluster was ignoring the unknown-to-it signature ConfigMaps.
+
+To mitigate these issues:
+
+* New versions of `oc` should continue to create particular configuration manifests types (e.g. release image signature ConfigMaps), even if newer clusters grow support for alternative mechanisms.
+* When pushing directly to a target cluster, `oc` might query the cluster to extract version information and warn about potential incompatibilies (e.g. "this cluster's version operator is too old to understand the release image signature ConfigMap I'm pushing" or, less usefully, "I'm a lot older than this cluster, and maybe a newer `oc` would be creating more manifests than I am").
+    The downsides to this approach are:
+    * It may be difficult to reliably determine if cluster components support a given object or not.
+    * There's no way to implement it in [the *pushing to disk* flow](#pushing-to-disk).
+* New versions of `oc` could create [custom resources][custom-resources] instead of using generic types such as ConfigMaps.
+    This way, configuration manifest application would generate errors when applying manifests to clusters which were too old to understand them.
+    But it is too late to use this approach for release image signatures, because the ConfigMap support has [already landed][cvo-config-map-signatures].
```

Comment:
We generally, in practice, accept then risk of version skew by instructing users some functions may not be available in older versions.  While there are downsides, they have never risen to the level of concern to justify expensive server interaction.

Reply:
> We generally, in practice, accept then risk of version skew by instructing users some functions may not be available in older versions.

This is covered [above][1] in my:

> Users in that situation might be surprised that the cluster was unable to locate release image signatures, not realizing that the older cluster was ignoring the unknown-to-it signature ConfigMaps.

line, right?  Also, is just "using a custom resource" an expensive server interaction?  Even if you push a ConfigMap, you're still connecting to the cluster and performing some server-side operations.

[1]: https://github.com/openshift/enhancements/commit/367e32de2e8ca0beb2116910d4038f0c65215f95#diff-c5f7837102728f1e74d092853f804d18R214

Reply:
Talking to the server is an expensive (in dev time) operation.  Basically we have sufficient experience that it's not worth it.

Reply:
Even with ConfigMap it's pretty straight forward, the cluster works with the items it knows about ignore everything above. Iow. if CM has fields signatures and somefancyfeature, older cluster will only work with signatures ignoring the other, whereas newer will work fine with both. That has proven fine in the past. I agree with Clayton here that client-side checking for version didn't work well. In CLIs we prefer to apply the current and fallback if that's not known to the current cluster.